library/modules/EventManager.cpp

```diff
@@ -527,12 +544,31 @@ static void manageJobCompletedEvent(color_ostream& out) {
     int32_t tick0 = eventLastTick[EventType::JOB_COMPLETED];
     int32_t tick1 = df::global::world->frame_counter;
 
-    multimap<Plugin*,EventHandler> copy(handlers[EventType::JOB_COMPLETED].begin(), handlers[EventType::JOB_COMPLETED].end());
-    unordered_map<int32_t, df::job*> nowJobs;
-    for ( df::job_list_link* link = &df::global::world->jobs.list; link != nullptr; link = link->next ) {
-        if ( link->item == nullptr )
-            continue;
-        nowJobs.emplace(link->item->id, link->item);
+    multimap<Plugin*, EventHandler> copy(handlers[EventType::JOB_COMPLETED].begin(), handlers[EventType::JOB_COMPLETED].end());
+    std::vector<JobCompleteData> nowJobs;
+    // predict the size in advance, this will prevent or reduce memory reallocation.
+    nowJobs.reserve(prevJobs.size());
+    for (const auto jobPtr : df::global::world->jobs.list) {
+        auto& job = *jobPtr;
+
+        auto seenIt = seenJobs.find(job.id);
+        if (seenIt != seenJobs.end()) {
+            df::job& seenJob = *seenIt->second;
+
+            //out.print("SEEN0: %d, %ld, %ld, %d, %ld, %d %s\n", seenJob.id, seenJob.items.size(), seenJob.job_items.elements.size(), seenJob.flags.bits.working, seenJob.general_refs.size(), getWorkerID(&seenJob), ENUM_ATTR(job_type, caption, seenJob.job_type));
+            //out.print("SEEN1: %d, %ld, %ld, %d, %ld, %d %s\n", job.id, job.items.size(), job.job_items.elements.size(), job.flags.bits.working, job.general_refs.size(), getWorkerID(&job), ENUM_ATTR(job_type, caption, job.job_type));
+
+            // The key here is to strategically check the most important bits to reduce churn.
+            if (seenJob.flags.whole != job.flags.whole
+                    || seenJob.items.size() != job.items.size()
+                    || seenJob.general_refs.size() != job.general_refs.size()) {
+                // Can't use emplace() for preexisting keys.
+                seenJobs[job.id] = JobUniquePtr(Job::cloneJobStruct(&job, true));
+            }
```

Comment:
ruh roh. I think that is UB.

seenJob is reference to pointer, but that pointer may be smashed by the assignment from seenJobs[job.id] = JobUniquePtr();  

No warnings, of course.